### ISSUE TITLE
QC check sample via CLI and in Workflow

### ISSUE THREAD

**Issue by hassanfa** [Clinical-Genomics/BALSAMIC#504]
QC check script added in PR#401 

The next step is to add this feature to:

### CLI: `balsamic report qc` (easy/medium effort)

To add it to BALSAMIC cli, it will be quite straightforward:
1. add a subcommand to `balsamic report`: `qc`
2. `balsamic report qc` will accept a case/sample config: `balsamic report qc --sample-config angry_bird.json`
3. It will run `qc_check` script internally, and print results to output.

### Snakemake workflow (large effort)
A QC rule with Python support and Jupyter Notebook template.

1. Create a Jupyter notebook
2. User Papermill (https://github.com/nteract/papermill) to parametrize the workflow
3. Create a Snakemake rule that reads the config file and populates the notebook and returns a PDF and Jupyter notebook
Example:

```
rule my_rule:
 input:
  notebook = "qc_notebook.ipynb",
  case_config = "angry_bird.json"

 output: "output.ipynb"
 run:
  import papermill as pm
  pm.execute_notebook(
     input.notebook,
     output,
     parameters = dict(case_config = input.case_config)
  )
```

To do above:
- Add papermill to `requirements.txt`

**Comment by Mropat** [comment]
I can add this to cg cli if you would like

**Comment by hassanfa** [comment]
@Mropat Lovely. Keyvan already picked up the `balsamic report qc`, and I will do the snakemake part. Could we please ask your help when it is done in balsamic? 

like, before storing calls balsmaic report qc to make sure exit 1 or 0. as simple as that. Does that work with cg? or do you prefer a concrete value?

**Comment by Mropat** [comment]
@hassanfa no this works fine!

**Comment by hassanfa** [comment]
cool! I'll tag you in PR for feedback then.

**Comment by hassanfa** [comment]
@keyvanelhami so you in the code: https://github.com/Clinical-Genomics/BALSAMIC/blob/develop/BALSAMIC/utils/qc_check.py The function `get_qc_check` is the main function and `failed_qc` checks if a QC failed.

In the workflow:

https://github.com/Clinical-Genomics/BALSAMIC/blob/c50641f8a540643cb76eea1bec02515f600c6895/BALSAMIC/workflows/balsamic.smk#L366-L375

1. MultiQC's parsed JSON files can be accessed using: `qc_dir + "multiqc_data/multiqc_picard_HSMetrics.json"`
1. `import failed_qc` or `import get_qc_check` in `rule all`, below `import datetime`
1. Send `multiqc_data/multiqc_picard_HSMetrics.json` and all the necessary input to `failed_qc` or `get_qc_check`.
1. `failed_qc` or `get_qc_check` should raise an error when QC parameters don't pass

**Comment by hassanfa** [comment]
@keyvanelhami this is an example and proof of concept of the implementation solution:


## Snakefile

This can be a workflow. Name this `Snakefile`:

```Python

# vim: syntax=python tabstop=4 expandtab
# coding: utf-8

from pathlib import Path

def read_qc_file():
  """ reads a qc file """
  with open("fold_80.txt", "r") as f:
    fold_80 = f.read()
  return float(fold_80)
        

def check_qc(var):
  """ checks qc and returns True or False"""
  
  if var > 1.8:
    return False
  else:
    return True


rule all:
  output: "analysis_finish"
  run:
    
    # Read and check QC
    sample_fold_80 = read_qc_file()
    result_ok = check_qc(var=sample_fold_80)

    # If QC failed raise an Error
    if not result_ok:
      print("Analysis QC failed")
      raise 
  
    # Create "analysis_finish" file
    Path(output[0]).touch()

```

## Fold_80 example file:

name this `fold_80.txt`
```
2.1
```

## Run:

`snakemake --cores 1`

**Comment by ivadym** [comment]
It will be addressed in #707.
